pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,213 @@
+		f.Close()
+	})
+
+	return ns, nil
+}
+
+const SO_NETNS_COOKIE = 71
+
+// GetNetNSCookie tries to retrieve the cookie of the host netns.
+func GetNetNSCookie() (uint64, error) {
+	s, err := unix.Socket(unix.AF_INET, unix.SOCK_STREAM, 0)
+	if err != nil {
+		return 0, err
+	}
+	defer unix.Close(s)
+
+	cookie, err := unix.GetsockoptUint64(s, unix.SOL_SOCKET, SO_NETNS_COOKIE)
+	if err != nil {
+		return 0, err
+	}
+
+	return cookie, nil
+}
+
+// FD returns the underlying file descriptor of the netNS
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNS) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all
+// references to it (including the one being closed here) are gone.
+func (h *NetNS) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func in the netns without changing the calling thread's netns.
+//
+// The code in f and any code called by f must NOT call [runtime.LockOSThread],
+// as this could leave the goroutine created by Do permanently pinned to an OS thread.
+func (h *NetNS) Do(f func() error) error {
+	g := new(errgroup.Group)
+
+	// Start the func in a new goroutine and lock it to an exclusive thread. This ensures that if
+	// execution of the goroutine fails unexpectedly before we call UnlockOSThread, the go runtime
+	// will ensure the underlying OS thread is disposed of, rather than reused in a potentially
+	// undefined state.
+	//
+	// See also: https://pkg.go.dev/runtime#UnlockOSThread
+	g.Go(func() error {
+		// save a handle to current network namespace
+		// and lock the thread - this will force this goroutine to be
+		// scheduled in a different thread, so we should be safe to mutate the netns
+		restore, err := saveRestoreCurrentThreadNetNS()
+		if err != nil {
+			return err
+		}
+		// unconditionally restore the snapshotted netns on termination, no matter what
+		defer restore()
+		unix.Setns(h.fd, unix.CLONE_NEWNET)
+		return f()
+	})
+
+	return g.Wait()
+}
+
+// saveRestoreCurrentThreadNetNS captures the current netns of the active thread
+// (whatever that is), and returns a func (which callers should `defer`) and an error if
+// for some reason the current netns is unobtainable.
+//
+// The returned func will restore the thread back to the captured netns it was running
+// under when this function was called, and unlock the current OS thread.
+//
+// If the netns of the current thread has not changed, this is effectively a no-op.
+func saveRestoreCurrentThreadNetNS() (func(), error) {
+	runtime.LockOSThread()
+	orig, err := getCurrent()
+	if err != nil {
+		// flaking out early, so unlock current thread
+		runtime.UnlockOSThread()
+		return nil, fmt.Errorf("get current namespace: %w", err)
+	}
+
+	return func() {
```

Comment:
```suggestion
	return func() error {
```
Let's make this function fallible so we can catch and log all errors.